datafusion/functions-aggregate/src/min_max/min_max_bytes.rs

```diff
@@ -0,0 +1,290 @@
+
+/// Implements Min/Max accumulators for "bytes" types ([`StringArray`], [`BinaryArray`], etc)
+///
+/// This implementation is general and stores the minimum/maximum for each groups in a byte array.
+///
+/// Note that for StringArray, there are potentially more efficient
+/// implementations by managing the string data buffer directly, but then garbage
+/// collection, memory management, and building the final array becomes more
+/// complex. See discussion on <https://github.com/apache/datafusion/issues/6906>
+#[derive(Debug)]
+pub(crate) struct MinMaxBytesAccumulator<F>
+where
+    F: FnMut(&str, &str) -> bool,
+{
+    /// The minimum/maximum value for each group
+    min_max: Vec<Option<String>>,
+    // todo use null state
+    // /// Have we seen any non-null values yet?
+    // null_state: NullState,
+    /// The data type of the array
+    data_type: DataType,
+    /// The total bytes of the string data (for pre-allocating the final array, and tracking memory usage)
+    total_data_bytes: usize,
+    /// The comparison function to use, called like `cmp(new_val, existing_val)`
+    ///
+    /// returns true if the `new_val` should replace `existing_val`
+    cmp: F,
+}
+
+#[derive(Debug, Clone, Copy)]
+enum MinMaxLocation<'a> {
+    /// the min/max value is stored in the existing `min_max` array
+    ExistingMinMax,
+    /// the min/max value is stored in the input array at the given index
+    Input(&'a str),
+}
+
+/// Implement the MinMaxBytesAccumulator with a comparison function
+/// for comparing strings
+impl<F> MinMaxBytesAccumulator<F>
+where
+    F: FnMut(&str, &str) -> bool,
+{
+    /// Create a new MinMaxBytesAccumulator
+    ///
+    /// # Arguments:
+    /// * `data_type`: The data type of the arrays that will be passed to this accumulator
+    /// * `cmp`: The comparison function to use. Called like `cmp(new_val, existing_val)`
+    ///   returns true if the `new_val` should replace `existing_val`.
+    pub fn new(data_type: DataType, cmp: F) -> Self {
+        Self {
+            min_max: vec![],
+            data_type,
+            total_data_bytes: 0,
+            cmp,
+        }
+    }
+
+    /// Set the specified group to the given value, updating memory usage appropriately
+    fn set_value(&mut self, group_index: usize, new_val: &str) {
+        match self.min_max[group_index].as_mut() {
+            None => {
+                self.min_max[group_index] = Some(new_val.to_string());
+                self.total_data_bytes += new_val.len();
+            }
+            Some(existing_val) => {
+                // Avoid re-allocating
+                //
+                // TODO check if this actually saves an allocation / if we
+                // actually avoid an allocation by using a mutable reference to
+                // the existing value)
+                self.total_data_bytes -= existing_val.len();
+                self.total_data_bytes += new_val.len();
+                *existing_val = String::from(new_val)
```

Comment:
I think this does an allocation?

We can `.clear()` the existing string and push to the existing string so it only allocate if the new string doesn't fit.

Reply:
Thank you -- I will try it